Refactor PE so fusion specializations are configurable

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72077
* #72036
* #72038
* #72040
* #72032
* #71651
* __->__ #71650
* #71173
* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.

Differential Revision: [D33801501](https://our.internmc.facebook.com/intern/diff/D33801501)

Commit:
Refactor PE so fusion specializations are configurable

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.h:
```diff
@@ -4,6 +4,17 @@
+
+using FusionStrategy = std::vector<std::pair<FusionBehavior, size_t>>;
+// E.g. {(Static, 2), (Dynamic, 10)}
+// Fuse with static shapes twice then fallback to 10 dynamic fusions, then stop
```
Why would we want multiple dynamic fusions? What is the use-case for that?

Krovatkin replied:
+1

eellison replied:
Even with "Dynamic" fusions, we are still specializing on rank, strides, dtype, device, etc. So if you think about a globally defined activation, there might end up with a number of dynamic specializations

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -68,6 +68,24 @@ static std::atomic<bool> executor_mode{true};
 #endif
 
+std::mutex fusion_strategy_lock;
+static FusionStrategy fusion_strategy = {{FusionBehavior::STATIC, 20}};
```
nit: would be great if we could add some documentation to python bindings that explain to a user how strategies work exactly and in which order they apply?

eellison replied:
My one worry is that we’re exposing implementation details… but yea I’ll add something with warnings/disclaimers at least

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -68,6 +68,24 @@ static std::atomic<bool> executor_mode{true};
+static FusionStrategy fusion_strategy = {{FusionBehavior::STATIC, 20}};
+
+FusionStrategy getFusionStrategy() {
+  fusion_strategy_lock.lock();
```
could you please `lock_guard` or something similar ?

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -68,6 +68,24 @@ static std::atomic<bool> executor_mode{true};
 static std::atomic<bool> profiling_mode{true};
 #endif
 
+std::mutex fusion_strategy_lock;
```
why isn't this static as well or in an anonymous namespace?

Krovatkin commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -752,10 +753,31 @@ void initJITBindings(PyObject* module) {
+            TORCH_WARN("Use _jit_set_fusion_strategy, bailout depth is deprecated. Setting to (Dynamic, ", depth, ")");
             size_t old_depth = getBailoutDepth();
-            getBailoutDepth() = depth;
+            FusionStrategy strat = {{FusionBehavior::DYNAMIC, depth}};
```
this will be changing the default behaviour where `_jit_set_bailout_depth` will only set `STATIC` strategies. Are we okay with this?

eellison replied:
sure let me update

Krovatkin commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -752,10 +753,31 @@ void initJITBindings(PyObject* module) {
             return old_depth;
           })
+      .def("_jit_set_fusion_strategy",
+          [](std::vector<std::pair<std::string, size_t>> strategy) {
```
should we consider adding bindings to the `STATIC` , `DYNAMIC` enum? It would simplify this code significantly and also remove the whole class of issues where a user mistypes in python `STAIC`.

eellison replied:
the typical thing in pytorch is to use strings as enum, and IMO this is an easier api

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -68,6 +68,24 @@ static std::atomic<bool> executor_mode{true};
+}
+
+FusionStrategy setFusionStrategy(FusionStrategy& strategy) {
+  fusion_strategy_lock.lock();
```
could we use a `lock_guard`?

Krovatkin commented on test/test_jit_fuser_te.py:
```diff
@@ -1877,6 +1876,23 @@ def eager(x, y):
                 g = torch.jit.last_executed_optimized_graph()
                 FileCheck().check("TensorExpr").run(g)
 
+    def test_exhaust_specializations(self):
```
nit: is there a test where we use multiple strategies? would be great if we could add one.

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -343,7 +361,21 @@ void runPreAutodiffPassPipeline(std::shared_ptr<Graph>& graph) {
+  }
+  // should never get here
+  TORCH_WARN("Stratgy changed mid-invocation, NYI");
+  return FusionBehavior::DYNAMIC;
```
we are changing the default behaviour here. I'd like to confirm we are okay with that.

eellison replied:
This should never get invoked, it means someone ran the fuser for a number of specializations, and then updated the global settings to a smaller number of allowed specializations.. I'm not sure what the behavior should be here. I'll change to STATIC conservatively

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.


[ghstack-poisoned]

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.


[ghstack-poisoned]

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.


[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33801501).

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.

Differential Revision: [D33801501](https://our.internmc.facebook.com/intern/diff/D33801501)

[ghstack-poisoned]

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.

Differential Revision: [D33801501](https://our.internmc.facebook.com/intern/diff/D33801501)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33801501).

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.

Differential Revision: [D33801501](https://our.internmc.facebook.com/intern/diff/D33801501)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33801501).

Commit:
Update on "Refactor PE so fusion specializations are configurable"

* 

Refactors PE so there is a current fusion strategy set, which will take in a vector of e.g. [(STATIC, 2), (DYNAMIC, 10)] which means fuse two static invocations then fuse 10 dynamic ones, then stop specializing.

Differential Revision: [D33801501](https://our.internmc.facebook.com/intern/diff/D33801501)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33801501).
